### PR TITLE
[HttpClient][Contracts] add HttpClient\ApiClientInterface et al.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR adds a new `ApiClientInterface` on top of #30413.
It's goal it to provide great DX when using (json) APIs.

It also provides an implementation of the interface in the `HttpClient` component:

image

### CODE REVIEW

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+     * Responses SHOULD be lazy, but their status code MUST be
+     * checked even if none of their public methods are called.
+     *
+     * @param string   $url     The URL of the resource
+     * @param iterable $options The options of the request
+     *
+     * @throws TransportExceptionInterface When an unsupported option is passed or
+     *                                     if any error happens at the transport layer
+     */
+    public function request(string $method, string $url, iterable $options = []): ResponseInterface;
```
Comment (dbu) [inline src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: in my opinion, the interface is a very leaky abstraction. rather than having all these things configured on a per-request basis, they should be configured on the client instance. having the options here tells people to leak all kind of details into their code interacting with the http client, instead of having that managed via DI on the http client itself.

if, say, there is one host where we do not want ssl verification because of an outdated certificate, i'd rather have a meta client that choses the correctly configured client based on a request detail like the hostname than littering this decision over all my code that interacts with the http client.

this is the reasoning we took for the PSR-18 request method, and if we would do the same here it would be possible to write an adapter from the symfony http client to use any psr-18 client. (the adapter in this PR is for using the symfony client as PSR-18 client, which is the other direction)

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Gets the response body as a string.
+     */
+    public function getContent(): string;
+
+    /**
+     * Returns info coming from the transport layer.
+     *
+     * This method SHOULD NOT throw any ExceptionInterface and SHOULD be non-blocking.
```
Comment (dbu) [inline src/Symfony/Contracts/HttpClient/ResponseInterface.php]: what is the behaviour if a non-blocking http request has been done and the information is not yet ready?

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: well, headers and body also appear as options in that signature for instance

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     *  - error - the error message when the transfer was aborted, null otherwise
+     *  - data - the value of the "data" request option, null if not set
+     *  - url - the last effective URL of the request
+     *
+     * When the "capture_peer_cert_chain" option is true, the "peer_certificate_chain"
+     * attribute SHOULD list the peer certificates as an array of OpenSSL X.509 resources.
+     *
+     * Other info SHOULD be named after curl_getinfo()'s associative return value.
+     */
+    public function getInfo(string $type = null);
```
Comment (stof) [inline src/Symfony/Contracts/HttpClient/ResponseInterface.php]: missing return type, either in the signature or the phpdoc.

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (dbu) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: yes, i am not arguing against headers and body, but against all the nitty-gritty http protocol options.

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: added in #30413

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: The behavior is that the method should return the current state of the request/response:
e.g; `0` as http_code when it did arrive yet, `primary_ip` empty or not before/after DNS resolution, etc. That's how both curl and native clients behave also (and what the contracts imply).

Comment (dbu) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: okay, that makes sense. maybe this could be said explicitly here? and the list below could explicitly state the not-yet-known values for all options, as it seems to not always be `null`. e.g. would raw_headers be an empty array or null? start_time?

File: src/Symfony/Component/HttpClient/CurlHttpClient.php [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            if ('GET' === $method && !$options['body'] && $expectedHeaders === $pushedHeaders) {
+                // Reinitialize the pushed response with request's options
+                return $pushedResponse->__construct($this->multi, $url, $options);
+            }
+        }
+
+        $curlopts = [
+            CURLOPT_URL => $url,
+            CURLOPT_DNS_USE_GLOBAL_CACHE => false,
+            CURLOPT_USERAGENT => 'Symfony HttpClient/Curl',
```
Comment (aheimburger) [inline src/Symfony/Component/HttpClient/CurlHttpClient.php]: This cannot be passed as an option ?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/CurlHttpClient.php]: Not directly as an option, but it can be defined as a header, and the header will have precedence.

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: hi @dbu, I have a different opinion here: all concrete HTTP clients have to deal with these settings. The defined options *do* abstract these. In PSR-18, you decided to not expose anything about the transport layer. This effectively forbids from controlling these settings in any interoperable ways. Your SSL example is a really good one: there is no way with PSR-18 itself to implement the logic you describe. I made a different choice and decided to expose the layer via the abstraction provided by the options. This will allow implementing interoperable decorators that do e.g. conditional SSL routing, and much more. IMHO, users need power tools, we should not restrict what they can do. If they use an HTTP client, it's legit that they can decide to control every aspect of it. Trying to save them from themselves could lead to the opposite effect: they will hack around the limitations and bound themselves to one implementation. That's exactly what happened to the Blackfire Player: its implementation is now bound to Guzzle because it needs DSN pre-resolution for security reasons; escaping the interop promise.

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: I read the comment again and it's clear to me... :)

Comment (dbu) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: its nice when its clear to you. but you know what to expect. when i looked at it, it was not 100% clear to me. and there will be people reading this with less of an idea of async than me, so i think it would be better to be fully explicit. add something like:

`The list below describes what value is returned when the information is not available yet.`

(and btw, i'd say "SHOULD NOT throw any exception" - or is it fine to throw some other exception that is not the ExceptionInterface? and, what happens when requesting an info that the client does not know about / did not implement?)

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (dbu) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: i agree that being overly limiting is a problem. however, this is not good flexibility but prompting bad patterns. having an $options parameter means that the configuration of how you want your client to work is carried throughout your application or hardcoded in all places that send requests. 
using correctly configured clients is a much better way to be sure all requests the application does are done with the correct settings. my example with ssl was to illustrate that even if you have a complicated situation, you can still inject one single client to your code and handle things in the client. the common use case however, will be that you send requests to one single server and want the same settings for all requests.

if i understand correctly, the list of options is not closed to what is mentioned in the interface, but could also be additional things. clients will not implement all of them, and some of the options are likely not well defined. if everything in the end uses lib_curl, you will have that as a de-facto standard, but clients that do not build on curl will likely not respect that.

Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: both ways are and should be supported: default options allow injecting preconfigured clients - of course, that's out of scope as no contracts can cover DI configuration.
Enabling decorators to precisely control HTTP-transport settings in an interoperable way is a goal of the proposed contracts. The list of options is in the contracts precisely because I want to make it a requirement to support them all - *or throw*. **All** options are supported by both native and curl implementations. There is zero binding to curl here. Note that *not* supporting an option is also possible, but then a client MUST throw. That applies to custom options of course so that if you rely on one of them, you'll bind yourself to one specific implementation, but at least you'll know when swapping the implementation breaks your need (very alike using methods that are only on implementations and not on some interfaces.)

File: src/Symfony/Component/HttpClient/Response/ApiResponse.php [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        if ($this->timeout) {
+            throw new TransportException('API response reached the inactivity timeout.');
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public static function complete(HttpClientInterface $client, iterable $apiResponses, ?float $timeout)
```
Comment (HeahDude) [inline src/Symfony/Component/HttpClient/Response/ApiResponse.php]: `: \Generator`?

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+            $this->info['raw_headers'][] = $h;
+        }
+
+        if (!$this->statusCode) {
+            throw new TransportException('Invalid or missing HTTP status line.');
+        }
+    }
+
+    private function checkStatusCode()
```
Comment (HeahDude) [inline src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: `: void`?

File: src/Symfony/Component/HttpClient/ApiClient.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        if ($responses instanceof ApiResponse) {
+            $responses = [$responses];
+        } elseif (!\is_iterable($responses)) {
+            throw new \TypeError(sprintf('%s() expects parameter 1 to be iterable of ApiResponse objects, %s given.', __METHOD__, \is_object($responses) ? \get_class($responses) : \gettype($responses)));
+        }
+
+        return new ApiResponseStream(ApiResponse::complete($this->client, $responses, $timeout));
+    }
+
+    private function requestApi(string $method, string $url, iterable $options): ApiResponseInterface
```
Comment (HeahDude) [inline src/Symfony/Component/HttpClient/ApiClient.php]: `: ApiResponse`? The method is private.

File: src/Symfony/Component/HttpClient/ApiClient.php [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        $options['buffer'] = true;
+        [, $options] = self::prepareRequest(null, null, $options, $this->defaultOptions);
+
+        return new ApiResponse($this->client->request($method, $url, $options));
+    }
+
+    private static function createInvalidArgumentException(string $message, string $file, int $line): \InvalidArgumentException
```
Comment (HeahDude) [inline src/Symfony/Component/HttpClient/ApiClient.php]: `: InvalidArgumentException`?

File: src/Symfony/Component/HttpClient/HttpClient.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+     * @param int   $maxHostConnections The maximum number of connections to a single host
+     */
+    public static function create(array $defaultOptions = [], int $maxHostConnections = 6): HttpClientInterface
+    {
+        if (\extension_loaded('curl')) {
+            return new CurlHttpClient($defaultOptions, $maxHostConnections);
+        }
+
+        return new NativeHttpClient($defaultOptions, $maxHostConnections);
+    }
```
Comment (HeahDude) [inline src/Symfony/Component/HttpClient/HttpClient.php]: What do you think of
```php
/**
      * @param array $defaultOptions     Default requests' options
      * @param int   $maxHostConnections The maximum number of connections to a single host
      */
    public static function createApi(array $defaultOptions = [], int $maxHostConnections = 6): ApiClientInterface
    {
        return new ApiClient(static::create($defaultOptions, $maxHostConnections));
    }
```
?

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (dbu) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: > The list of options is in the contracts precisely because I want to make it a requirement to support them all - or throw.

i either missed that part or its not yet explicitly written out. when going with this approach, i think its a very good point, because it makes things a lot more reliable. 
do i get it correctly that the list is open, so additional options might be supported, but clients MUST throw an exception if they don't know any option that they have been passed?

i am still not convinced that this is the right approach. but with that clarification, the whole thing looks a lot better to me :+1:

Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php]: My communication style is never explicit enough, thanks for helping me brain-dump :) I'm sure there is room for clarifications by PRs after merging. You're right about added and unsupported options, that's what I meant by the description next to `@throws`.

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: I added some related sentences. PR welcome after merge if you're up to. Thanks for asking.

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: visual debt :)

File: src/Symfony/Component/HttpClient/HttpClient.php (hunk 9 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/HttpClient.php]: Not sure it's useful. But can be discussed and done later in another PR!

File: src/Symfony/Component/HttpClient/ApiClient.php (hunk 8 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/ApiClient.php]: method removed

File: src/Symfony/Component/HttpClient/Response/ApiResponse.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: done

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php]: I added this:
`Implementations are not required to support all options described above; they can also support more custom options; but in any case, they MUST throw a TransportExceptionInterface when an unsupported option is passed.`

**Comment by nicolas-grekas** [comment]
PR rebased, ready for reviews, votes, merge :)

File: src/Symfony/Contracts/HttpClient/Test/ApiClientTestCase.php [hunk 10]
```diff
@@ -0,0 +1,647 @@
...
+    public function testGzipBroken()
+    {
+        $client = $this->getApiClient();
+        $response = $client->get('http://localhost:8057/gzip-broken');
+
+        $this->expectException(TransportExceptionInterface::class);
+        $response->getContent();
+    }
+
+    private static function createInvalidArgumentException(string $message, string $file, int $line): \InvalidArgumentException
```
Comment (HeahDude) [inline src/Symfony/Contracts/HttpClient/Test/ApiClientTestCase.php]: Is it used?

File: src/Symfony/Component/HttpClient/HttpClient.php (hunk 9 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/HttpClient.php]: OK, added as `ApiClient::create()` :)

File: src/Symfony/Component/HttpClient/Response/ApiResponse.php [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+                // Skip timed out responses but throw on destruct if unchecked
+                $r->timeout = $chunk->isTimeout();
+
+                if ($r->timeout || !$chunk->isLast()) {
+                    continue;
+                }
+            } catch (TransportExceptionInterface $e) {
+                // Ensure errors are always thrown, last resort on destruct
+                $r->error = $e;
+                $e = null;
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: "Ensure PHP can clear memory as early as possible" added on top of the line

File: src/Symfony/Contracts/HttpClient/Test/ApiClientTestCase.php (hunk 10 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/Test/ApiClientTestCase.php]: removed

File: src/Symfony/Component/HttpClient/Response/ApiResponse.php [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+    /**
+     * @internal
+     */
+    public static function complete(HttpClientInterface $client, iterable $apiResponses, ?float $timeout): \Generator
+    {
+        $responses = new \SplObjectStorage();
+
+        foreach ($apiResponses as $r) {
+            $responses[$r->response] = $r;
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: Sure, documented, and now check with proper exception. Only ApiResponse are accepted because that's the design of the component: the responses are not standalone object, they're bound by "friendship" with the client that emitted them. That's how clients know how to stream of complete them.

[the remark this answers is not in the document]
Comment (ostrolucky) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: That seems like break of contract of `ApiClientInterface`. `ApiClientInterface::complete` says you can use `ApiResponseInterface`, but `ApiClient::complete` rejects everything that is not `ApiResponse`.

[the remark this answers is not in the document]
Comment (HeahDude) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: The contract is on `ApiClientInterface::complete` not `ApiResponseInterface::complete`

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: It's very alike PSR 6 cache items: they are not interoperable, like responses are not interoperable from one client to another.

Here is how PSR 6 tells it:

>  Calling Libraries SHOULD NOT assume that an Item created by one Implementing Library is compatible with a Pool from another Implementing Library.

Same for responses.

[the remark this answers is not in the document]
Comment (ostrolucky) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: I wonder why not design fitting contract? We are in design phase. Contract should declare everything implementations need. If even main implementation works around the contract, that's not a good signal.

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: But a contract is not only about type hints. If the contracts says that responses cannot be passed from one client implementation to another, that's a the contracts, and a perfectly valid one. Exactly as PSR 6.

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: Rereading, I may not have answered your question @ostrolucky.

> Contract should declare everything implementations need

That's actually what happens here. Tell me if I'm wrong, but I understand from your comment that you think we should look for completely standalone response classes: we should design each interface as a standalone contract (eg responses) that could we reused independently with any other interface in the domain (eg clients). What I'm trying to explain above is that a class is not the only possible unit to define the boundaries of reusable code units. Here, like in PSR6, the client is the only "legal" factory for response objects. For this reason, we do not need to define any contracts to make them talk to each other: that communication is purely internal details.

Another way to look at the problem is that one: do we need to work on a set of interfaces that would allow any response objects to work with any clients? Let's be more concrete: do we need to define interfaces that would allow a response created by a client based on **curl** to be streamed by a client based on **fopen**?

Maybe it's possible, by adding more abstractions in between. But which problem would this solve in practice? Nobody will ever need to do that.

So, in conclusion, the client+response form a tuple that are bound together by some internal detail the outside world should not care about. That's actually the design decision that makes the contracts work so seamlessly and so consistently.

File: src/Symfony/Component/HttpClient/ApiClient.php [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        return new self(HttpClient::create($defaultOptions, $maxHostConnections));
+    }
+
+    public function __construct(HttpClientInterface $client, array $defaultOptions = [])
+    {
+        $this->client = $client;
+
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, $this->defaultOptions);
```
Comment (dbu) [inline src/Symfony/Component/HttpClient/ApiClient.php]: would it be possible to refactor the HttpClientTrait to have separate methods for the url and the options? i see that one section of the trait prepareRequest uses null === $url as a flag to do auth or not. i think this could be refactored too, to be more explicit and to have an options resolution that is not preparing the request.

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+namespace Symfony\Contracts\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\Test\ApiClientTestCase;
+
+/**
+ * Provides flexible methods for interacting with HTTP APIs.
+ *
+ * Unless otherwise specified, implementations MUST send a JSON-compatible
+ * "accept" header - typically "application/json".
```
Comment (dbu) [inline src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: Unless specified otherwise in the 'headers' $option, implementations MUST ...

Comment (dbu) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: actually, why? if the client must be for json, why is there an interface? are there more meaningful implementations than json_decode?

what if i want to implement this for an xml api? (not soap, there array would make little sense as a response type, but some generic xml thing)

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+interface ApiClientInterface
+{
+    const OPTIONS_DEFAULTS = HttpClientInterface::OPTIONS_DEFAULTS;
+
+    /**
+     * Requests an HTTP resource using the GET verb.
+     *
+     * @throws TransportExceptionInterface When an unsupported option is passed
+     */
+    public function get(string $url, iterable $options = []): ApiResponseInterface;
```
Comment (dbu) [inline src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: these http method named methods would make sense with any client, not only for an API client. why is this tied into api client and not a decorator for the plain client?

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * Unless otherwise specified, implementations MUST send a JSON-compatible
+ * "accept" header - typically "application/json".
+ *
+ * @see HttpClientInterface for a description of options and how they MUST work
+ * @see ApiClientTestCase   for a reference test suite
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface ApiClientInterface
```
Comment (dbu) [inline src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: imho it would be useful to also have the generic `request` method on this interface, for APIs that accept special http methods.

**Review by dbu — COMMENTED** [review]
so essentially, these contracts are for automatically converting the response body to array, right?

is that worth all the interfaces? what if we instead had a general http client interface with methods and tell people to use json_decode on the response body?

(this is meant as an actual question - i am fine with an answer that says "yes, there is a significant benefit" - in that case i'd love some additional phpdoc to explain what the benefit is, why people should use this)

File: src/Symfony/Component/HttpClient/ApiClient.php (hunk 13 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/ApiClient.php]: That's not completely possible as there is some coupling between options and url - the most obvious example is the `base_uri` options. I'd be happy to review a PR if you want to work on one.

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php (hunk 14 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: Except `toArray()`, which doesn't fit other content types, the interface fits xml yes.
The idea here is to tell ppl that if they don't explicitly an `Accept` header, then they can expect application/json to be sent for them. Then, if one client is configured to send xml, one should ensure that such a client would be injected only where that default is expected.
Would that make sense? How would you formulate that?

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php (hunk 15 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: What I mean by "api" is that the response bodies are small enough to be buffered in memory. E.g. a `.iso` of a video file wouldn't fit this definition - one should use `HttpClientInterface` then.
A decorator wouldn't provide the required abstraction level: I'm looking for something ppl can type-hint for without binding them to one implementation.

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php (hunk 16 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: We already have such a thing: it's `HttpClientInterface`

**Comment by nicolas-grekas** [comment]
> so essentially, these contracts are for automatically converting the response body to array, right?

I'm actually wondering if we shouldn't move `toArray()` to the base `ResponseInterface`... so, not really :)

These contracts are for talking with endpoints that serve reasonably small responses - json/xml/text payloads.
Provided responses are reasonably small, the `complete()` method becomes a useful tool: no need to deal with chunks anymore, you'll only need to care about complete and buffered responses.

The goal is also to provide those get/post/put/etc method that covers 99% of API endpoint needs.
So, this aim at providing a useful abstraction to access APIs with reduced boilerplate.

File: src/Symfony/Component/HttpClient/Response/ApiResponse.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (ostrolucky) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: > contracts says that responses cannot be passed from one client implementation to another

1. Can you point me to a part of the contract in http-client which says this? I don't see it.
1. Even if contract says that, how does this non-interoperability benefit user?

> That's actually the design decision that makes the contracts work so seamlessly and so consistently.

My impression is it makes whole thing work less seamlessly and less consistently. Even you had to workaround your own contract with SplObjectStorage, because http client implementation rejects any request it didn't itself create. Anybody creating their own implementation will have to jump through same hoops. And it has code smells like methods in response accepting other responses and task scheduling bult in, instead of in client or separate class. This is less likely to happen with contracts defining boundaries properly, like having ResponseInterface::isFinished

> Exactly as PSR 6.

PSR-6 doesn't forbid anybody to make pools work with items from other pools. It's unfortunate that symfony/cache chose this option. symfony/http-client goes even more overboard and doesn't even reuse same Response (Item) class across Clients (ItemPools). Imagine how much more work you would have if every cache adapter came with its own item class. This is happening here, just nobody realizes because there isn't many implementations (yet).


> that communication is purely internal details

This is leaky internal detail and will surface when trying to implement own Response class. And contract doesn't warn even in text form. Not even clients do that. Every client pretends it accepts interface responses, while deeper layer that it uses rejects everything else than concrete, final class.

> do we need to define interfaces that would allow a response created by a client based on curl to be streamed by a client based on fopen?

Http-client goes way more far than that. It's not that consumer is not able to use curl response in fopen client, but they are not able to use ANY other response. They cannot even use wrapped curl response in curl client, or curl response subclass. If consumer wants to create new Response, they must create new client as well. That includes cases when consumer would wish to add just one utility method. Combined with the fact everything has to be decorator because nothing is inheritable, this looks too verbose. And when user begins to have a need to have own response classes, they can no longer safely use interface, because anything else than their own implementation might blow up when passed to `ApiClientInterface::complete` or `HttpClientInterface::stream`, better to typehint to their own client instead across whole application to avoid can of worms. Even then, user must be careful each time passing these to external libraries. Maybe I am overreacting because this applies only to those mentioned bulk methods (and their use will be rare), but this will be PITA when somebody will need to use those in this context.

> But which problem would this solve in practice? Nobody will ever need to do that.

I might wish to create: XmlResponse, CsvResponse, StripeResponse, StringeableResponse, MockResponse, UserCreateResponse, TransactionAwareResponse, TraceableResponse etc. I might wish to use unique response class for every API endpoint. Use case this solves is preserve ability to safely pass these to `ApiClientInterface::complete` or `HttpClientInterface::stream` without unwinding & rewinding these response wrappers (which cannot even happen without instanceof checks, because they can be passed to different layers)

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ApiResponse.php]: Relates to https://github.com/symfony/symfony/pull/30413#issuecomment-468828899

File: src/Symfony/Component/HttpClient/ApiClient.php [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            throw new \TypeError(sprintf('%s() expects parameter 1 to be iterable of ApiResponse objects, %s given.', __METHOD__, \is_object($responses) ? \get_class($responses) : \gettype($responses)));
+        }
+
+        return new ApiResponseStream(ApiResponse::complete($this->client, $responses, $timeout));
+    }
+
+    private function requestApi(string $method, string $url, iterable $options): ApiResponse
+    {
+        $options['buffer'] = true;
+        [, $options] = self::prepareRequest(null, null, $options, $this->defaultOptions);
```
Comment (ostrolucky) [inline src/Symfony/Component/HttpClient/ApiClient.php]: this method requires array, iterable passed

File: src/Symfony/Component/HttpClient/ApiClient.php [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public function options(string $url, iterable $options = []): ApiResponseInterface
+    {
+        return $this->requestApi('OPTIONS', $url, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function complete($responses, float $timeout = null): ApiResponseStreamInterface
```
Comment (ostrolucky) [inline src/Symfony/Component/HttpClient/ApiClient.php]: This should document it accepts `ApiResponse[]` only

File: src/Symfony/Component/HttpClient/ApiClient.php (hunk 17 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/ApiClient.php]: Thanks. All `iterable` type hints where leftovers, now fixed.

File: src/Symfony/Component/HttpClient/ApiClient.php (hunk 18 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/ApiClient.php]: added

**Comment by nicolas-grekas** [comment]
PR rebased on top of #30499 - waiting for it to be merged.
`ApiResponseInterface` is gone now.

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php (hunk 14 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: Reformulated.Bette?

**Comment by nicolas-grekas** [comment]
Rebased, ready.

**Comment by nicolas-grekas** [comment]
Thinking twice here, I'm fine if we don't merge. With #30499 in, talking with JSON endpoint might already be easy enough.

The benefit of this new interface would be that:
- it provides helper methods for standard HTTP verbs;
- it provides the `complete()` method, that prevents dealing with chunks directly, by yielding only completed responses.

The downside is that it requires ppl to make a choice between this interface and `HttpClientInterface` - and we know making a choice is hard.

I nobody comments to rescue this PR, I'm going to close it in a few days :)

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php (hunk 16 above)
Comment (dbu) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: but you do not extend the HttpClientInterface here. there is a very small cost to also proxy that base method (assuming that the api client will implement such a method anyways)

File: src/Symfony/Contracts/HttpClient/ApiClientInterface.php (hunk 14 above)
Comment (dbu) [reply src/Symfony/Contracts/HttpClient/ApiClientInterface.php]: yes!

**Comment by dbu** [comment]
> it provides helper methods for standard HTTP verbs;

you could provide a HttpMethodsClient (or whatever you want to call it) that decorates a HttpClientInterface with methods for the HTTP verbs.

and apart from that, i share your concern that having 2 clients can be confusing for people. i am not sure of the added value. maybe adding something to the response to make it convenient for non-streaming situations would solve the issue with less additional code and interfaces?

**Comment by fabpot** [comment]
I also think that this PR does not bring enough value and makes things more complex for our users as they need to choose which abstraction to use. Let's close.
